ex2_hal/iris/hardware_interface/source/iris.c

```diff
@@ -0,0 +1,460 @@
+#include "logger.h"
+
+/*
+ * Optimization points
+ * - Full error coverage is desired
+ */
+
+
+/**
+ * @brief
+ *   Sends take a picture command to Iris
+ *
+ * @return
+ *   Returns IRIS_HAL_OK if equipment handler returns IRIS_ACK, else IRIS_HAL_ERROR
+ **/
+IrisHALReturn iris_take_pic() {
+    IrisLowLevelReturn ret;
+
+    controller_state = SEND_COMMAND;
+
+    while (controller_state != FINISH && controller_state != ERROR_STATE) {
+        switch (controller_state) {
+            case SEND_COMMAND:
+            {
+                ret = send_command(IRIS_TAKE_PIC);
+                if (ret == IRIS_ACK) {
+                    controller_state = FINISH;
+                } else {
+                    controller_state = ERROR_STATE;
+                }
+                break;
+            }
+            case FINISH:
+            {
+                ex2_log("Iris returns ACK on take a picture command");
+                return IRIS_HAL_OK;
+            }
+            case ERROR_STATE:
+            {
+                ex2_log("Iris returns NACK on take a picture command");
+                return IRIS_HAL_ERROR;
+            }
+        }
+    }
+    return IRIS_HAL_ERROR;
+}
+
+/**
+ * @brief
+ *   Sends a get image length command to Iris, and expects to receive the
+ *   image length
+ *
+ * @return
+ *   Returns the length in bytes of image stored on Iris
+ **/
+IrisHALReturn iris_get_image_length(uint32_t *image_length) {
+    IrisLowLevelReturn ret;
+
+    controller_state = SEND_COMMAND;
+
+    while (controller_state != FINISH && controller_state != ERROR_STATE) {
+        switch (controller_state) {
+            case SEND_COMMAND:
+            {
+                ret = send_command(IRIS_GET_IMAGE_LENGTH);
+                if (ret == IRIS_ACK) {
+                    controller_state = GET_DATA;
+                } else {
+                    controller_state = FINISH;
+                }
+                break;
+            }
+            case GET_DATA:
+            {
+                uint16_t * image_length_buffer = (uint16_t*) pvPortMalloc(MAX_IMAGE_LENGTH * sizeof(uint16_t));
+                memset(image_length_buffer, 0, MAX_IMAGE_LENGTH);
+                if (image_length_buffer == NULL) {
+                    ex2_log("Failed attempt to dynamically allocate memory under iris get image length");
+                    return IRIS_HAL_ERROR;;
+                }
+                ret = get_data(image_length_buffer, MAX_IMAGE_LENGTH);
+                if (ret == IRIS_HAL_OK) {
+                    controller_state = FINISH;
+                } else {
+                    controller_state = ERROR_STATE;
+                }
+                /* It is expected that the first byte in the buffer will be the LSB */
+                *(image_length) = (uint32_t)((uint8_t)image_length_buffer[2]<<16 | (uint8_t)image_length_buffer[1]<<8 | (uint8_t)image_length_buffer[0]); // Concatenate image_length_buffer
+
+                free(image_length_buffer);
```

Comment:
Use vPortFree. (irrelevant if you allocate as a stack variable, like in my other comments)